src/back/wgsl/writer.rs

```diff
@@ -880,25 +880,47 @@ impl<W: Write> Writer<W> {
                 };
 
                 let l2 = level.next();
-                if !cases.is_empty() {
-                    for case in cases {
-                        match case.value {
-                            crate::SwitchValue::Integer(value) => {
-                                writeln!(self.out, "{}case {}{}: {{", l2, value, type_postfix)?;
+                let mut new_case = true;
```

Comment:
idea(non-blocking): Maybe we could change `SwitchCase` model to remove `fall_through` and make `body` an `Option`? Concretely:

```rust
/// <snip>
struct SwitchCase {
    // …

    /// When [`None`], a subsequent case should contain the body associated with this case.
    body: Option<Block>, // replaces `fall_through` and current definition of `body`
}
```

A one-off `enum` might also serve well here, but I'm less certain about that.

Reply:
The other frontends still make use of fall_through with non empty bodies though,

Reply:
> This code that decides when to write "case" is fine, but it does seem like the IR is not making the writer's life easy. We might be able to improve the IR, but that should be a separate PR.

I thought about this but I could not come up with a satisfactory change because we still need to support "real" fall-through for GLSL and SPV.

> Granted, this would need to be changed when we implement fallthrough, but I think if we trade `split_inclusive` for `.enumerate().filter()` and then gather up all the consecutive bodies we need to concatenate, we could keep this as a loop whose iterations correspond to emitted "case" statements, not IR cases, which is what I think makes it clearer.

I don't see how `.enumerate().filter()` would work since we still need to emit the case values of the `fall_through` cases. As far as I can tell the only way to do this would be to allocate a new structure and have one pass restructuring the data then another to write it which I don't think is really worth it.